Fixing Egardia alarm status shown as unknown after restart

jeroenterheerdt opened the pull request:
## Description:
In some situations, after a restart of HASS the Egardia alarm system was reported with status 'unknown'. This PR fixes this issue by doing an extra status update regardless of exact config of the pythonegardia component.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jeroenterheerdt committed:
Fixing Egardia alarm status shown as unknown afer restart

jeroenterheerdt committed:
Fixing Egardia alarm status shown as unknown afer restart

MartinHjelmare commented on homeassistant/components/alarm_control_panel/egardia.py:
```diff
@@ -162,8 +164,12 @@ def parsestatus(self, status):
     def update(self):
         """Update the alarm status."""
         if not self._rs_enabled:
```
Instead, move this line to the `should_poll` property that you can define and have it return `False` if `self._rs_enabled` is true otherwise return `True`.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/egardia.py:
```diff
@@ -103,6 +103,8 @@ def __init__(self, name, egardiasystem, hass, rs_enabled=False,
         else:
             self._rs_codes = rs_codes
 
+        self.getstatusfromsystem()
```
Remove this.

MartinHjelmare commented on homeassistant/components/alarm_control_panel/egardia.py:
```diff
@@ -162,8 +164,12 @@ def parsestatus(self, status):
-            self.parsestatus(status)
+            self.getstatusfromsystem()
+
+    def getstatusfromsystem(self):
```
Remove this.

MartinHjelmare reviewed (changes requested):
You want the `update` method to be called once before adding the entity. This is already done via the second parameter to `add_devices`. So the correct solution is to also have the `update` method do something for non polling entities. You also need to add a call to update home assistant state in the event callback.

On another note, I think you should enable the pythonegardia server to be run in a thread, accept a callback, and to be started from home assistant. That way you can have a better integration and the users will have an easier setup time.

jeroenterheerdt commented:
@MartinHjelmare thanks - I'll look into the review a little later. On your note regarding running egardiaserver in a thread, do you have any pointers on how to do this?

hudashot committed:
Export climate status and target temperature to Prometheus (#10919)

* Export climate metrics to Prometheus.

This adds climate_state and temperature_c metrics for each climate
device.

* Add more climate states to state_as_number

dasos committed:
Tado ignore invalid devices (#10927)

* Ignore devices without temperatures

* Typo

* Linting

* Removing return false

* Another typo. :(

* Spelling received correctly

MartinHjelmare commented:
You can eg look at [pymysensors `TCPGateway`](https://github.com/theolind/pymysensors/blob/master/mysensors/gateway_tcp.py) where we create an instance of a class,  which also subclasses a thread. This gateway is a socket client, not a server, but the principle is the same. This is part of the mysensors integration in home assistant. The pymysensors gateway also accepts a callback that is called by the gateway whenever a state changes on a device. This way the gateway can push updates to home assistant.

jeroenterheerdt commented:
@MartinHjelmare looking at your review I do not see how these code changes get the same result as mine. I need the status to be called immeditately regardless of whether or not the user uses the egardiaserver. With your proposal if the user uses egardiaserver (self._rs_enabled = true) then self._egardiasystem.getstate() is never called. That is exactly what I needed to fix using this code.

molobrakos committed:
Upgrade tellduslive library version (closes https://github.com/home-assistant/home-assistant/issues/10922) (#10950)

molobrakos committed:
Upgrade tellduslive library version (closes https://github.com/home-assistant/home-assistant/issues/10922) (#10950)

PaulAnnekov committed:
don't ignore voltage data if sensor data changed (#10925)

MartinHjelmare commented:
The `update` method where `self._egardiasystem.getstate()` is called, before your change in this PR, will be called during addition of the entity regardless of `self._rs_enabled`. To not have the `update` method return due to `self._rs_enabled` being true, you should move that line to a `should_poll` property that you define in the entity.

Do you follow?

MartinHjelmare commented:
Eg:
```py
def update(self):
    """Update the alarm status."""
    self.getstatusfromsystem()

@property
def should_poll(self):
    """Poll if no report server is enabled."""
    if not self._rs_enabled:
        return True
    return False
```

mateuszdrab committed:
Fix linksys_ap.py by inheriting DeviceScanner (#10947)

As per issue #8638, the class wasn't inheriting from DeviceScanner, this commit patches it up.

mateuszdrab committed:
Fix linksys_ap.py by inheriting DeviceScanner (#10947)

As per issue #8638, the class wasn't inheriting from DeviceScanner, this commit patches it up.

Commit:
Add ADS component (#10142)

* add ads hub, light and switch

* add binary sensor prototype

* switch: use adsvar for connection

* fix some issues with binary sensor

* fix binary sensor

* fix all platforms

* use latest pyads

* fixed error with multiple binary sensors

* add sensor

* add ads sensor

* clean up after shutdown

* ads component with platforms switch, binary_sensor, light, sensor

add locking

poll sensors at startup

update state of ads switch and light

update ads requirements

remove update() from constructors on ads platforms

omit ads coverage

ads catch read error when polling

* add ads service

* add default settings for use_notify and poll_interval

* fix too long line

* Fix style issues

* no pydocstyle errors

* Send and receive native brightness data to ADS device to prevent issues with math.floor reducing brightness -1 at every switch

* Enable non dimmable lights

* remove setting of self._state in switch

* remove polling

* Revert "remove polling"

This reverts commit 7da420f82385a4a5c66a929af7025c00ed197e86.

* add service schema, add links to documentation

* fix naming, cleanup

* re-remove polling

* use async_added_to_hass for setup of callbacks

* fix comment.

* add callbacks for changed values

* use async_add_job for creating device notifications

* set should_poll to False for all platforms

* change should_poll to property

* add service description to services.yaml

* add for brigthness not being None

* put ads component in package

* Remove whitespace

* omit ads package

craigjmidwinter committed:
Reload closest store on api menu request (#10962)

* reload closest store on api request

* revert change from debugging

craigjmidwinter committed:
Reload closest store on api menu request (#10962)

* reload closest store on api request

* revert change from debugging

HerrHofrat committed:
Gearbest sensor (#10556)

* Added Gearbest Sensor

* Updated required files

* Fixed houndci-bout findings

* Fix tox lint errors

* Changed code according to review
Implemented library version 1.0.5

* Fixed houndci-bot findings

* Fixed tox lint issues

* Updated item schema to use has_at_least_one_key
Added conf constants

* Remove CONF_ constants and import them from homeassistant.const

* Removed CurrencyConverter from hass
Fixed couple of issues found by MartinHjelmare

b10m committed:
Add Ziggo Mediabox XL media_player (#10514)

* Add Ziggo Mediabox XL media_player

* Using pypi module ziggo-mediabox-xl now.

* Code review changes

ziotibia81 committed:
Generic thermostat initial_operation_mode (#10690)

* Generic thermostat restore operation mode

* Test restore operation mode

* Fix trailing whitespace

* Fix line too long

* Fix test duplicate entity_id

* Fix test

* async_added_to_hass modify modify internal state

* Test inital_operation_mode

* More restore state tests

* Fix whitespace

* fix test_custom_setup_param

* Test "None" target temp

emlove committed:
Use new build path for dev translations (#10937)

masarliev committed:
Add option to set default hide if away for new devices (#10762)

* Option to change hide_if_away

* tests fix

* change new device defaults

* tests and requested changes

* fix assert

Commit:
Allow chime to work for wink siren/chime (#10961)

* Allow Wink siren/chimes to work

* Updated requirements_all.txt

Commit:
Allow chime to work for wink siren/chime (#10961)

* Allow Wink siren/chimes to work

* Updated requirements_all.txt

balloob committed:
Revert pychromecast update (#10989)

* Revert pychromecast update

* Update cast.py

balloob committed:
Revert pychromecast update (#10989)

* Revert pychromecast update

* Update cast.py

andrey-git committed:
Require FF43 for latest js (#10941)

* Require FF43 for latest js

`Array.prototype.includes` added in Firefox 43

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

* Update __init__.py

andrey-git committed:
Require FF43 for latest js (#10941)

* Require FF43 for latest js

`Array.prototype.includes` added in Firefox 43

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

* Update __init__.py

balloob committed:
Version bump to 0.59.2

balloob committed:
Merge pull request #10990 from home-assistant/release-0-59-2

0.59.2

balloob committed:
Merge branch 'master' into dev

DanNixon committed:
Allow disabling the LEDs on TP-Link smart plugs (#10980)

balloob committed:
Update frontend to 20171206.0

masarliev committed:
Meraki AP Device tracker (#10971)

* Device tracker for meraki AP

* styles fix

* fix again

* again

* and again :)

* fix hide if away

* docs and optimization

* tests and fixes

* styles

* styles

* styles

* styles

* styles fix. Hope last

* clear track new

* changes

* fix accuracy error and requested changes

* remove meraki from .coveragerc

* tests and minor changes

* remove location

jeroenterheerdt commented:
@MartinHjelmare I think your review is not functionally the same as my code. I need the self._egardiasystem.getstate() method to be called regardless of the value of self._rs_enabled. The flow is supposed to be like this: 1) call self._egardiasystem.getstate() 2) if self._rs_enabled = true subscribe to the system_status event. if self._rs_enabled = false we need to call self._egardiasystem.getstate() when the update() method is called.

MartinHjelmare commented:
My code example will do what you outline above. And also fix the bug of missing state update for callbacks.

MartinHjelmare commented:
See here:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alarm_control_panel/__init__.py#L140-L149

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_component.py#L214-L220

pvizeli committed:
Update tradfri.py (#10991)

jeroenterheerdt commented:
@MartinHjelmare I finally get it now - I did not understand the should_poll property - I see now that is part of the alarm control panel framework for HASS. Thanks.

OddBloke committed:
webostv: Ensure source exists before use (#10959)

In a case where either (a) an incorrect source name is used, or (b) the
TV isn't currently queryable (e.g. it's off), we get tracebacks because
we assume the source that we are being asked to select exists in
self._source_list.

This makes the lookup code a little more rugged, and adds in a warning
message (in place of the current exception).

Commit:
Ensure Docker script files uses LF line endings to support Docker for Windows. (#10067)

alanfischer committed:
Added Vera scenes (#10424)

* Added Vera scenes

* Fixed flake8 issues

* Fixed comments

* Moved vera to use hass.data

* Made requested changes

jeroenterheerdt committed:
Merge branch 'egardia' of https://github.com/jeroenterheerdt/home-assistant into egardia

jeroenterheerdt committed:
Updating egardia.py based on review

homeassistant commented:
Hi @stlehmann, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10940).

Once you do that we will be able to review and accept this pull request.

Thanks!

jeroenterheerdt commented:
@MartinHjelmare I made the changes - it works fine - let me know if it is OK. I will look into improving the egardiaserver (reportserver) installation better a little later.

MartinHjelmare commented:
Your branch contains old unrelated commits now. You can try to rebase your commits onto fresh dev.

jeroenterheerdt commented:
I tried that, but it did not help. Should I just close this commit and re-create a new one?

MartinHjelmare commented:
Yes, that works too.

jeroenterheerdt commented:
@MartinHjelmare any chance you could help me out developing the gateway you mentioned? I do not know where to start. Currently, the users have to run an egardiaserver component and as far as I understood we can get rid of that by implementing something like a gateway. The egardiaserver component is only used for passing along status info from the alarm control panel. The control panel sends its status to a configurable IP/port. The egardiaserver component runs as a listener on that IP/port and sends an event to HASS. Can you help me write code that removes the need to egardiaserver? I need some help getting started... thanks.

MartinHjelmare commented:
@jeroenterheerdt I'll have a look. Maybe we can continue the discussion on discord? Are you there?

jeroenterheerdt commented:
@MartinHjelmare yeah, I am there as well. Let's talk.
